forte/data/extractor/attribute_extractor.py

```diff
@@ -0,0 +1,134 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Dict, Any, Union, Iterable
+from collections import abc
+from ft.onto.base_ontology import Entry, Annotation
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.data.converter.feature import Feature
+from forte.data.extractor.base_extractor import BaseExtractor
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "AttributeExtractor"
+]
+
+
+class AttributeExtractor(BaseExtractor):
+    r"""AttributeExtractor extracts feature from the attribute of entry.
+
+    Args:
+        config: An instance of `Dict` or
+            :class:`forte.common.configuration.Config`
+
+            attribute: str. Required. The attribute name of the
+                entry from which features will be extracted. For
+                example, "text" attribute of Token.
+    """
+    def __init__(self, config: Union[Dict, Config]):
+        super().__init__(config)
+
+        if "attribute" not in self.config:
+            raise AttributeError("attribute needs to be specified in "
+                                "the configuration of an AttributeExtractor.")
+
+    @property
+    def state(self) -> Dict:
+        state = super().state
+        state.update({
+            "attribute": self.config.attribute
+        })
+        return state
+
+    @classmethod
+    def from_state(cls, state: Dict) \
+        -> Union["BaseExtractor", "AttributeExtractor"]:
+        obj = BaseExtractor.from_state(state)
+        additional = {
+            "attribute": state["attribute"]
+        }
+        obj.config = Config(obj.config, additional,
+                            allow_new_hparam=True)
+        obj.__class__ = cls
+        return obj
+
+    @staticmethod
+    def get_attribute(entry: Entry, attr: str) -> Any:
+        r"""Functionality: Get the attribute from entry. You can
+        overwrite this function if you have sepcial way to get the
+        attribute from entry.
+        """
+        return getattr(entry, attr)
+
+    @staticmethod
+    def set_attribute(entry: Entry, attr: str, value: Any):
+        r"""Functionality: Set the attribute of an entry to value.
+        You can overwrite this function if you have sepcial way to
+        set the attribute.
+        """
+        if attr == "text":
+            raise AttributeError("text attribute of entry cannot "
+                                "be changed.")
+        setattr(entry, attr, value)
+
+    def update_vocab(self, pack: DataPack, instance: Annotation):
+        r"""Functionality: Get all attributes of one instance and
+        add them into the vocabulary.
+        """
+        for entry in pack.get(self.config.entry_type, instance):
+            self.add(self.get_attribute(entry, self.config.attribute))
```

Comment:
Vocabulary need to take `Hashable` but `get_attribute` is `Any`. So what happens if we extract something that is not hashable?

To solve this problem, we should think about whether everything need to be placed into vocabulary, can something be skipped from the vocab? And how can users specify that?

Maybe one way to do this, is when the attribute is skipped from vocabulary, we may return the raw value?

Reply:
1. Right we need to make sure the element added to vocabulary is Hashable type. So I think we need to make the assumption here or do an assertion here that if the user call update_vocab, they should be aware that they are using hashable attribute.
2. For the case, get_attribute return unhashable element. I think the user is better to set "vocal_method" to be "raw", so that there will not be vocabulary and update_vocab should not be called.

Reply:
https://github.com/asyml/forte/blob/900cf93880e83178a62c4d183f264268e3ea44c6/forte/data/extractor/base_extractor.py#L89

Reply:
https://github.com/asyml/forte/blob/900cf93880e83178a62c4d183f264268e3ea44c6/forte/data/extractor/base_extractor.py#L172